File: 658-FindKClosestElements/658-FindKClosestElements.py
"""

658. Find K Closest Elements

Given a sorted integer array arr, two integers k and x, return the k closest integers to x in the array. The result should also be sorted in ascending order.

An integer a is closer to x than an integer b if:

    |a - x| < |b - x|, or
    |a - x| == |b - x| and a < b

Example 1:

Input: arr = [1,2,3,4,5], k = 4, x = 3
Output: [1,2,3,4]

Example 2:

Input: arr = [1,2,3,4,5], k = 4, x = -1
Output: [1,2,3,4]

Constraints:

    1 <= k <= arr.length
    1 <= arr.length <= 104
    arr is sorted in ascending order.
    -104 <= arr[i], x <= 104

"""

class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        heap = []
        for a in arr:
            heapq.heappush(heap, (- abs(x - a), -a))
            if len(heap) > k : heapq.heappop(heap)
        res = [-heapq.heappop(heap)[1] for _ in range(k)]
        return sorted(res)

File: 658-FindKClosestElements/658-FindKClosestElements.java
/*

658. Find K Closest Elements
Medium

Given a sorted array, two integers k and x, find the k closest elements to x in the array. The result should also be sorted in ascending order. If there is a tie, the smaller elements are always preferred.

Example 1:

Input: [1,2,3,4,5], k=4, x=3
Output: [1,2,3,4]

Example 2:

Input: [1,2,3,4,5], k=4, x=-1
Output: [1,2,3,4]

Note:

    The value k is positive and will always be smaller than the length of the sorted array.
    Length of the given array is positive and will not exceed 104
    Absolute value of elements in the array and x will not exceed 104

*/

class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        // create a max heap to return min values;
        List<Integer> resList = new ArrayList<>();
        Queue<Integer> heap = new PriorityQueue<Integer>((n1, n2) -> Math.abs(x - n2) == Math.abs(x - n1) ? n2 - n1 :  Math.abs(x - n2) - Math.abs(x - n1));
        for(int a : arr){
            heap.add(a);
            if(heap.size() > k)
                heap.poll();
        }
        while(!heap.isEmpty())
            resList.add(heap.poll());
        Collections.sort(resList);
        return resList;
    }
}

class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> resList = new ArrayList<>();
        int left = 0, right = arr.length - k;
        while (left < right) {
            int mid = (left + right) / 2;
            if (x - arr[mid] > arr[mid + k] - x)
                left = mid + 1;
            else
                right = mid;
        }
        for(int i = left; i < left + k; i++)
            resList.add(arr[i]);
        return resList;
    }
}
